crates/db_connection_pool/src/dbconnection/odbcconn.rs

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::datatypes::SchemaRef;
+use arrow::record_batch::RecordBatch;
+use arrow_odbc::arrow_schema_from;
+use arrow_odbc::OdbcReaderBuilder;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::physical_plan::memory::MemoryStream;
+use datafusion::sql::TableReference;
+use futures::lock::Mutex;
+use odbc_api::handles::Statement;
+use odbc_api::handles::StatementImpl;
+use odbc_api::parameter::InputParameter;
+use odbc_api::CursorImpl;
+use snafu::prelude::*;
+use snafu::Snafu;
+
+use crate::DbConnectionPool;
+
+use super::AsyncDbConnection;
+use super::DbConnection;
+use super::Result;
+use odbc_api::Connection;
+
+pub trait ODBCSyncParameter: InputParameter + Sync {
+    fn as_input_parameter(&self) -> &dyn InputParameter;
+}
+
+impl<T: InputParameter + Sync> ODBCSyncParameter for T {
+    fn as_input_parameter(&self) -> &dyn InputParameter {
+        self
+    }
+}
+
+pub type ODBCParameter = Box<dyn ODBCSyncParameter>;
+pub type ODBCDbConnection<'a> = (dyn DbConnection<Connection<'a>, ODBCParameter>);
+pub type ODBCDbConnectionPool<'a> =
+    dyn DbConnectionPool<Connection<'a>, ODBCParameter> + Sync + Send;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    ArrowODBCError { source: arrow_odbc::Error },
+    ODBCAPIError { source: odbc_api::Error },
+}
+
+pub struct ODBCConnection<'a> {
+    pub conn: Arc<Mutex<Connection<'a>>>,
+}
+
+impl<'a> DbConnection<Connection<'a>, ODBCParameter> for ODBCConnection<'a>
+where
+    'a: 'static,
```

Comment:
Why is `'static` necessary here?

Reply:
More FFI fun! 😄 

A [connection handle](https://docs.rs/odbc-api/7.0.0/src/odbc_api/handles/connection.rs.html#41) closes over an [environment handle](https://docs.rs/odbc-api/7.0.0/src/odbc_api/handles/environment.rs.html#32) which ends up being a static ref. The ODBC [environment](https://docs.rs/odbc-api/7.0.0/odbc_api/struct.Environment.html) is global and there must be one per process.

Our structs can (and still are) dropped just fine (as they themselves are not static). You can validate this by e.g. doing something like this and triggering a reconnection (fsnotify on saving the dataset seems to do the trick):

```rust
impl<'a> std::ops::Drop for ODBCConnection<'a> {
    fn drop(&mut self) {
        println!("weeeee i'm gettin dropped!");
    }
}
```

Reply:
Thanks for detailed response. Good to know.